### PR TITLE
Comment additions

### PR DESCRIPTION
**Pull request by rsafi082**
General comment additions where seen fit, some of the comments need to
be finished at the end of fetch_data.py

### CODE REVIEW

File: active_learning/admin.py [hunk 1]
```diff
@@ -2,4 +2,5 @@
 
 from .models import Article
 
+# register Article model with admin site
```
Comment (SeanLF) [inline active_learning/admin.py]: remove

File: active_learning/apps.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+# imports
```
Comment (SeanLF) [inline active_learning/apps.py]: too obvious, not necessary

File: active_learning/fetch_data.py [hunk 3]
```diff
@@ -11,7 +11,18 @@
 # for feature_selection
 from sklearn.feature_selection import SelectKBest, chi2
 
-
+'''
```
Comment (SeanLF) [inline active_learning/fetch_data.py]: please keep 2 empty lines before a method definition

Comment (SeanLF) [reply active_learning/fetch_data.py]: In python, the convention seems to be to put docstrings inside a function. See [here](https://www.python.org/dev/peps/pep-0257/)

File: active_learning/models.py [hunk 4]
```diff
@@ -1,5 +1,9 @@
+# imports
 from django.db import models
 
+'''
```
Comment (SeanLF) [inline active_learning/models.py]: remove

**Review by SeanLF — CHANGES_REQUESTED** [review]
- Remove `imports` comments
- Move class and method comment blocks into their definition (see relevant comment for link to explanation)
- in files without a class, keep 2 blank lines above `def method_name(...`

File: active_learning/fetch_data.py [hunk 5]
```diff
@@ -49,32 +82,57 @@ def download_articles(articles):
...
 
     # return the stemmed tokens
     return stems
 
 
 def tf_idf_matrix(raw_documents, tokenizing_method):
+    '''
+    Uses the raw documents with a specific tokenizing method in order to create and return a rizer and matrix
+
+    Parameters
```
Comment (SeanLF) [inline active_learning/fetch_data.py]: missing param comments

File: active_learning/fetch_data.py [hunk 6]
```diff
@@ -84,6 +142,21 @@ def tf_idf_matrix(raw_documents, tokenizing_method):
 
 
 def feature_selection(tf_idf_matrix, document_classes_for_matrix, max_features, feature_names):
+    '''
+    Uses a tf idf matrix as well as the given document classes for the matrix in order to extract an array of features to use
+
+    Parameters
```
Comment (SeanLF) [inline active_learning/fetch_data.py]: missing param comments

File: active_learning/models.py [hunk 7]
```diff
@@ -1,5 +1,8 @@
...
```
Comment (SeanLF) [inline active_learning/models.py]: remove please

File: active_learning/our_labeler.py [hunk 8]
```diff
@@ -3,11 +3,37 @@
 from active_learning.utils import format_pusher_channel_name, get_pusher_client
 from os import environ
 
+'''
```
Comment (SeanLF) [inline active_learning/our_labeler.py]: move inside of class

File: active_learning/our_labeler.py [hunk 9]
```diff
@@ -18,10 +44,22 @@ def __init__(self, **kwargs):
         pusher.connection.bind('pusher:connection_established', self.connect_handler)
         pusher.connect()
 
-    # Feature is a dictionary with url and id as keys
+
```
Comment (SeanLF) [inline active_learning/our_labeler.py]: don't need two spaces for methods inside of a class

**Review by SeanLF — CHANGES_REQUESTED** [review]

File: active_learning/fetch_data.py (hunk 6 above)
Comment (rsafi082) [inline active_learning/fetch_data.py]: This one and the one above will have to be filled in later, I wasn't quite sure how to add the parameter comments in as of now.

File: active_learning/models.py (hunk 7 above)
Comment (rsafi082) [reply active_learning/models.py]: Resolved

File: active_learning/our_labeler.py (hunk 8 above)
Comment (rsafi082) [reply active_learning/our_labeler.py]: Resolved

File: active_learning/our_labeler.py (hunk 9 above)
Comment (rsafi082) [reply active_learning/our_labeler.py]: Resolved

File: active_learning/fetch_data.py [hunk 10]
```diff
@@ -25,17 +38,37 @@ def bing_news_search(query, max_results=100, offset=0):
...
 def get_links(search_query, max_results):
+    '''
+    Loop through the returned results and add the urls to a list
+
+    Parameters
+    ----------
+    search_query : string
+        The search query term to be used
+    max_results : integer
+        The number of results expected per page
```
Comment (SeanLF) [inline active_learning/fetch_data.py]: please fix: this is the final expected number of items to fetch

File: active_learning/fetch_data.py [hunk 11]
```diff
@@ -49,41 +82,81 @@ def download_articles(articles):
 
 
 def clean_string(string):
-    # Format the string to be case insensitive (lower case)
-    # Remove special formatting characters (ex: new line character)
-    # return the cleaned string
+    '''
+    Return the given string in lower case removing any newline characters
+
+    Parameters
```
Comment (SeanLF) [inline active_learning/fetch_data.py]: don't think this is necessary, you can remove this

File: active_learning/fetch_data.py [hunk 12]
```diff
@@ -49,41 +82,81 @@ def download_articles(articles):
...
     return string.lower().replace('\n', ' ')
 
 
 def tokenize_and_stem(string):
+    '''
+    Return the given string after splitting it up, word by word, then getting the stem of each work
+
+    Getting the stem of words allows for easier identification of a word in all of its forms
+
+    Parameters
```
Comment (SeanLF) [inline active_learning/fetch_data.py]: think you can remove the parameters section, obvious from the description

File: active_learning/fetch_data.py [hunk 13]
```diff
@@ -49,41 +82,81 @@ def download_articles(articles):
...
-        # if token not in self.stopwords:
         stems.append(stemmer.stem(token))
 
     # return the stemmed tokens
     return stems
 
 
 def tf_idf_matrix(raw_documents, tokenizing_method):
+    '''
+    Uses the raw documents with a specific tokenizing method in order to create and return a rizer and matrix
```
Comment (SeanLF) [inline active_learning/fetch_data.py]: typo: `rizer` to `vectorizer`

File: active_learning/learning.py [hunk 14]
```diff
@@ -42,12 +79,32 @@ def classify_svm(self, X, y):
         print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))
         return scores
 
+
     def learn(self, auto_label=True, active_learning_strategy=5, num_queries=50, train_size=0.005):
+        '''
+        Uses the specified learning strategy and then queries the oracle/learner before printing the accuracy at which the articles are labeled
```
Comment (SeanLF) [inline active_learning/learning.py]: remove

> /learner

File: active_learning/learning.py [hunk 15]
```diff
@@ -31,7 +54,21 @@ def __init__(self, **kwargs):
...
             self.y.append(article.class_label_id)
 
+
     def classify_svm(self, X, y):
+        '''
+        Uses SVM to classify articles before calculating and printing the accuracy
+
+        Parameters
+        ----------
+        self : self reference
```
Comment (SeanLF) [inline active_learning/learning.py]: remove this, as we don't actually pass a self reference when calling this function

File: active_learning/learning.py [hunk 16]
```diff
@@ -42,12 +79,32 @@ def classify_svm(self, X, y):
...
         return scores
 
+
     def learn(self, auto_label=True, active_learning_strategy=5, num_queries=50, train_size=0.005):
+        '''
+        Uses the specified learning strategy and then queries the oracle/learner before printing the accuracy at which the articles are labeled
+
+        Parameters
+        ----------
+        self : self reference
```
Comment (SeanLF) [inline active_learning/learning.py]: remove this, as we don't actually pass a self reference when calling this function

File: active_learning/learning.py [hunk 17]
```diff
@@ -42,12 +79,32 @@ def classify_svm(self, X, y):
...
+            An object reference to the object itself
+        auto_label : boolean
+            Boolean representing whether we want to use our own labeler or an ideal labeler
+        active_learning_strategy : integer
+            An integer representing the learning strategy to use
+        num_queries : integer
+            The number of queries for the oracle to label
+        train_size : integer
+            An integer representing the percentage of the corpus to use in the labeled set
```
Comment (SeanLF) [inline active_learning/learning.py]: real number between 0 and 1

File: active_learning/our_labeler.py [hunk 18]
```diff
@@ -3,11 +3,37 @@
 from active_learning.utils import format_pusher_channel_name, get_pusher_client
 from os import environ
 
-
 class OurLabeler(Labeler):
+    '''
```
Comment (SeanLF) [inline active_learning/our_labeler.py]: documentation for wrong class

File: active_learning/our_labeler.py [hunk 19]
```diff
@@ -3,11 +3,37 @@
...
+        A list of labels
+    '''
+
     lbl = None
 
     def __init__(self, **kwargs):
+        '''
+        Initializes the labeler by:
+
+        1. Getting a list of labels
```
Comment (SeanLF) [inline active_learning/our_labeler.py]: labels is an argument when instantiating this class

File: active_learning/our_labeler.py [hunk 20]
```diff
@@ -18,10 +44,21 @@ def __init__(self, **kwargs):
...
         pusher.connect()
 
-    # Feature is a dictionary with url and id as keys
     def label(self, feature):
+        '''
+        Return the label produced by the oracle
+
+        Parameters
+        ----------
+        self : self reference
```
Comment (SeanLF) [inline active_learning/our_labeler.py]: remove this, as we don't actually pass a self reference when calling this function

File: active_learning/our_labeler.py [hunk 21]
```diff
@@ -18,10 +44,21 @@ def __init__(self, **kwargs):
...
+            An object reference to the object itself
+        feature : dictionary
+            A dictionary where the keys are the ids and the urls are the values
```
Comment (SeanLF) [inline active_learning/our_labeler.py]: only two keys in this dictionary, singularize `ids` and `urls`

File: active_learning/utils.py [hunk 22]
```diff
@@ -13,6 +21,10 @@ def format_pusher_channel_name(channel_name):
 
 
 def get_pusher_client():
+    '''
+    Returns the pusher using environment for the pusher's attributes
```
Comment (SeanLF) [inline active_learning/utils.py]: pusher **client**
environment **variables**

File: active_learning/views.py [hunk 23]
```diff
@@ -16,35 +16,59 @@
...
+    '''
+
     presence_channel_name = format_pusher_channel_name(environ['PRESENCE_CHANNEL_NAME'])
     PUSHER_KEY = environ['PUSHER_KEY']
     return render(request, 'active_learning/iframe.html', {'presence_channel_name': presence_channel_name, 'PUSHER_KEY': PUSHER_KEY})
 
 
 def learn(request):
+    '''
+    After retrieving an oracle's response, learn from the enw data
```
Comment (SeanLF) [inline active_learning/views.py]: starts the active learning process which requires another tab to be loaded at the active_learning root (iframe) page for the oracle to label articles

File: active_learning/views.py [hunk 24]
```diff
@@ -16,35 +16,59 @@
...
+    '''
+
     do_learn.after_response()
     return HttpResponse(status=200)
 
 
 @after_response.enable
 def do_learn():
+    '''
+    Learn from any new data, when you're done, stop learning
```
Comment (SeanLF) [inline active_learning/views.py]: begins the active learning process asynchronously

File: active_learning/views.py [hunk 25]
```diff
@@ -16,35 +16,59 @@
...
+    '''
+
     learn = Learn()
     learn.learn(False)
 
 
 def get_articles(request):
+    '''
+    Get the list of articles after having queried and gotten a response
```
Comment (SeanLF) [inline active_learning/views.py]: fetch articles from bing

**Review by SeanLF — CHANGES_REQUESTED** [review]

**Comment by SeanLF** [comment]
I'm happy to accept with this pull request once you fix the last issue in `fetch_data.py` and undo your changes to the report.

**Review by SeanLF — APPROVED** [review]
Nice! 👍
